pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	configv1 "github.com/openshift/api/config/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	machineapi "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	machineconfigapi "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	upgradev1alpha1 "github.com/openshift/managed-upgrade-operator/pkg/apis/upgrade/v1alpha1"
+	operatorv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	once                sync.Once
+	upgrader            ClusterUpgrader
+	UpgradeStepOrdering = []upgradev1alpha1.UpgradeConditionType{
+		upgradev1alpha1.UpgradeValidated,
+		upgradev1alpha1.UpgradePreHealthCheck,
+		upgradev1alpha1.UpgradeScaleUpExtraNodes,
+		upgradev1alpha1.CommenceUpgrade,
+		upgradev1alpha1.ControlPlaneMaintWindow,
+		upgradev1alpha1.ControlPlaneUpgraded,
+		upgradev1alpha1.AllMastersUpgraded,
+		upgradev1alpha1.RemoveControlPlaneMaintWindow,
+		upgradev1alpha1.WorkersMaintWindow,
+		upgradev1alpha1.AllWorkerUpgraded,
+		upgradev1alpha1.PostUpgradeFix,
+		upgradev1alpha1.RemoveExtraScaledNodes,
+		upgradev1alpha1.UpdateSubscriptions,
+		upgradev1alpha1.PostUpgradeVerification,
+		upgradev1alpha1.RemoveMaintWindow,
+		upgradev1alpha1.PostClusterHealthCheck,
+	}
+)
+
+const (
+	TIMEOUT_SCALE_EXTRAL_NODES = 30 * time.Minute
+
+	LABEL_UPGRADE = "upgrade.managed.openshift.io"
+)
+
+type ClusterUpgrader map[upgradev1alpha1.UpgradeConditionType]UpgradeStep
+
+// Ordering returns the ordering of predicates.
+func Ordering() []upgradev1alpha1.UpgradeConditionType {
+	return UpgradeStepOrdering
+}
+
+type UpgradeStep func(client client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error)
+
+func NewUpgrader(reqLogger logr.Logger, client client.Client) ClusterUpgrader {
+	once.Do(func() {
+		upgrader = map[upgradev1alpha1.UpgradeConditionType]UpgradeStep{
+			upgradev1alpha1.UpgradeValidated:              ValidateUpgradeConfig,
+			upgradev1alpha1.UpgradePreHealthCheck:         PreClusterHealthCheck,
+			upgradev1alpha1.UpgradeScaleUpExtraNodes:      EnsureExtraUpgradeWorkers,
+			upgradev1alpha1.ControlPlaneMaintWindow:       CreateControlPlaneMaintWindow,
+			upgradev1alpha1.CommenceUpgrade:               CommenceUpgrade,
+			upgradev1alpha1.ControlPlaneUpgraded:          ControlPlaneUpgraded,
+			upgradev1alpha1.AllMastersUpgraded:            AllMastersUpgraded,
+			upgradev1alpha1.RemoveControlPlaneMaintWindow: RemoveControlPlaneMaintWindow,
+			upgradev1alpha1.WorkersMaintWindow:            CreateWorkerMaintWindow,
+			upgradev1alpha1.AllWorkerUpgraded:             AllWorkersUpgraded,
```

Comment:
Recommend we rename this to "AllWorkersUpgraded" (add "s" in Workers), so that it is consistent with the function name.